src/shader.h

```diff
@@ -74,6 +76,51 @@ class IShaderConstantSetter
 			bool is_highlevel) = 0;
 };
 
+
+class IShaderConstantSetterFactory
+{
+public:
+	virtual ~IShaderConstantSetterFactory() {};
+	virtual IShaderConstantSetter* create() = 0;
+};
+
+
+template <typename T, bool is_pixel, std::size_t count=1>
+class CachedShaderSetting {
+	const char *m_name;
+	T m_sent[count];
+	bool has_been_set;
+protected:
+	CachedShaderSetting(const char *name) : m_name(name), has_been_set(false) {}
```

Comment:
Why is name not a template parameter too?

In fact, why not just make everything in the template parameter a constructor parameter to eliminate the need for templates altogether?